0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu
+package core
+
+/**
+ * The four elements of any Iglu-compatible schema
+ * SchemaKey can be used both for Self-describing instances
+ * and Self-describing Schemas
+ */
+case class SchemaKey(
+  vendor: String,
+  name: String,
+  format: String,
+  version: SchemaVer) {
+
+  /**
+   * Converts a SchemaKey into a path which is compatible
+   * with most local and remote Iglu schema repositories.
+   *
+   * @return a path usable for addressing local and remote
+   *         Iglu schema lookups
+   */
+  def toPath: String =
+    s"$vendor/$name/$format/${version.asString}"
+
+  /**
+   * Converts the SchemaKey back to an Iglu-format schema URI
+   *
+   * @return the SchemaKey as a Iglu-format schema URI
+   */
+  def toSchemaUri: String =
+    s"iglu:${toPath}"
+}
+
+/**
+ * Companion object contains a custom constructor for
+ * an Iglu SchemaKey.
+ */
+object SchemaKey {
+
+  /**
+   * Canonical regular expression for SchemaKey
+   */
+  val schemaUriRegex = "^iglu:([a-zA-Z0-9-_.]+)/([a-zA-Z0-9-_]+)/([a-zA-Z0-9-_]+)/([0-9]+-[0-9]+-[0-9]+)$".r
+
+  /**
+   * Regular expression to extract SchemaKey from path
+   */
+  val schemaPathRegex = "^([a-zA-Z0-9-_.]+)/([a-zA-Z0-9-_]+)/([a-zA-Z0-9-_]+)/([0-9]+-[0-9]+-[0-9]+)$".r
+
+  /**
+   * Default [[Ordering]] instance for [[SchemaKey]]
+   * Sort keys alphabetically AND by ascending SchemaVer
+   * (so initial Schemas will be in the beginning)
+   *
+   * Usage:
+   * {{{
+   *   import com.snowplowanalytics.iglu.core.SchemaKey
+   *   implicit val schemaOrdering = SchemaKey.ordering
+   *   keys.sorted
+   * }}}
+   */
+  val ordering = Ordering.by { (key: SchemaKey) =>
+    (key.vendor, key.name, key.format, key.version.model, key.version.revision, key.version.addition)
+  }
```

Comment:
I like the idea of a schema ordering. We could also eventually have:

* A similar ordering but for SchemaVers instead of SchemaKeys
* A scala.math.PartialOrdering for SchemaKeys where if two schemas have a different vendor, name, or format, neither of them is less than the other.